### ISSUE TITLE
Embedding Optimization: Switch to Kanon-2 for Legal Documents (+3-5% accuracy)

### ISSUE THREAD

**Issue by michalprusek** [SUJBOT2/SUJBOT2#49]
## 🎯 Objective
Improve retrieval accuracy by **+3-5%** through switching to **Kanon-2** embeddings, the #1 ranked model on MLEB 2025 benchmark, specifically optimized for legal/technical documents.

## 📊 Current State (PHASE 4)
- **Default**: `text-embedding-3-large` (OpenAI, 3072D)
- **Local option**: `bge-m3` (1024D, free, GPU-accelerated on M1/M2/M3)
- **Already supported but not default**: `kanon-2` (Voyage AI) ✨
- Cost: $0.13 per 1M tokens (OpenAI)

## 🏆 SOTA 2025: MLEB Benchmark Results

| Rank | Model | Provider | NDCG@10 | Use Case |
|------|-------|----------|---------|----------|
| **#1** | **kanon-2** | Voyage AI | **86.0%** | **Legal/Finance** |
| #2 | voyage-3-large | Voyage AI | 85.7% | General |
| #8 | text-embedding-3-large | OpenAI | 82.3% | General |

**Kanon-2 advantages**:
- ✅ Optimized for legal and financial documents
- ✅ +4.5% better NDCG@10 vs. text-embedding-3-large
- ✅ Already implemented in codebase (just not default)
- ✅ Similar cost: $0.06 per 1M tokens

## 🔬 Research Foundation
1. **MLEB Benchmark** (2025):
   - Kanon-2 ranks #1 across 6 legal/finance datasets
   - Outperforms OpenAI by **+3-5% on legal retrieval tasks**

2. **Domain-specific embeddings** (2024 studies):
   - **Legal-BGE-base**: Fine-tuned BGE for legal text
   - **Legal-E5**: E5 model adapted for legal domain
   - Outperform general-purpose embeddings on legal benchmarks

3. **Voyage AI research**:
   - Kanon-2 uses legal corpus pre-training
   - Better semantic understanding of legal terminology
   - Superior handling of complex legal references

### Phase 1: Switch to Kanon-2 (3-5% improvement, 1-2 days)
- [ ] **Update configuration**:

```bash
# .env
EMBEDDING_MODEL=kanon-2
VOYAGE_API_KEY=your-voyage-api-key
```

- [ ] **Re-index all documents**:
  - Run: `uv run python run_pipeline.py data/regulace/`
  - Time: 1-2 hours for typical corpus
  - Note: All existing vector stores will be invalidated

- [ ] **Benchmark against current setup**:
  - Test on 30-query evaluation dataset
  - Compare Precision@5, Recall@10, NDCG@10
  - Measure retrieval quality on legal queries

**Expected improvement: +3-5% retrieval accuracy**

### Phase 2: Legal-Specific Fine-tuning (optional, advanced, +2-3%)
- [ ] **Fine-tune embeddings** on your specific legal corpus:
  - Use Legal-BGE-base as starting point
  - Fine-tune on Czech legal documents (if applicable)
  - Train with triplet loss on legal terminology pairs

```python
# File: scripts/fine_tune_embeddings.py (NEW)

from sentence_transformers import SentenceTransformer, InputExample, losses
from torch.utils.data import DataLoader

def fine_tune_legal_embeddings(
    base_model="BAAI/bge-base-en-v1.5",
    legal_corpus_path="data/legal_corpus.txt",
    output_path="models/legal-bge-czech"
):
    """
    Fine-tune BGE embeddings on Czech legal documents.
    
    Training data format:
    - Positive pairs: (legal term, definition)
    - Negative pairs: (legal term, unrelated text)
    """
    # Load base model
    model = SentenceTransformer(base_model)
    
    # Prepare training data
    train_examples = [
        InputExample(texts=["waste disposal", "management of hazardous materials"]),
        # ... more legal term pairs
    ]
    
    # Create DataLoader
    train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=16)
    
    # Define loss
    train_loss = losses.MultipleNegativesRankingLoss(model)
    
    # Train
    model.fit(
        train_objectives=[(train_dataloader, train_loss)],
        epochs=1,
        warmup_steps=100
    )
    
    # Save
    model.save(output_path)
```

- [ ] Deploy fine-tuned model as local embedding option
- [ ] Benchmark vs. kanon-2

**Expected additional improvement: +2-3%** (if corpus-specific)

### Phase 3: Hybrid Embedding Strategy (advanced, optional)
- [ ] **Multi-model ensemble**:
  - Use kanon-2 for semantic search (primary)
  - Use Legal-BGE for entity matching and cross-document resolution
  - Ensemble fusion with learned weights

- [ ] **Adaptive embedding selection**:
  - Simple queries → faster model (bge-m3)
  - Complex legal queries → kanon-2
  - Cross-document tasks → Legal-BGE

## 📈 Expected Impact

| Metric | Current (text-emb-3-large) | After Kanon-2 | Improvement |
|--------|---------------------------|---------------|-------------|
| **Precision@5** | ~75% | **~78-80%** | **+3-5%** |
| **Recall@10** | ~65% | **~68-71%** | **+3-6%** |
| **NDCG@10** | ~82% | **~86%** | **+4.5%** |
| **Cost per 1M tokens** | $0.13 | **$0.06** | **-54%** ↓ |
| **Legal term accuracy** | Baseline | **Better** | Domain-optimized |

### Pros ✅
- **Immediate improvement**: +3-5% retrieval accuracy
- **Lower cost**: $0.06 vs $0.13 per 1M tokens
- **Already supported**: Just change config, no code changes
- **Legal-optimized**: Better understanding of legal terminology

### Cons ⚠️
- **Re-indexing required**: All documents must be re-embedded (1-2 hours)
- **API dependency**: Requires Voyage API key
- **Dimension change**: Kanon-2 uses different dimensions than OpenAI

### Migration Path
1. Backup current vector stores
2. Update .env with kanon-2 config
3. Re-index documents in batch (use eco mode for cost savings)
4. Run evaluation to verify improvement
5. Deploy to production

## 🔧 Implementation Priority
**HIGH PRIORITY** - Low effort, high reward ("quick win")
- **Time investment**: 1-2 days (mostly re-indexing)
- **Expected ROI**: +3-5% accuracy + lower costs
- **Risk**: Low (easily reversible by changing config)

## 🧪 Evaluation Plan
Before/after comparison:
- [ ] Run evaluation on 30-query test dataset
- [ ] Compare Precision@5, Recall@10, NDCG@10
- [ ] Test on legal-specific queries (citations, clause types, regulations)
- [ ] Measure cost per query
- [ ] Document results in `evaluation/kanon2_migration.md`

## 📚 References
- MLEB Benchmark (2025): Kanon-2 #1 ranked model (86% NDCG@10)
- Voyage AI documentation: https://docs.voyageai.com/
- Legal-BGE research (2024): Domain-specific fine-tuning benefits
- BGE-M3 (BAAI 2024): Multi-lingual, multi-granularity embeddings

See `papers/cost_effectiveness.md` section 2.4 and `IMPROVEMENTS.md` section 2.4 for detailed research.
